lighthouse-core/test/audits/seo/tap-targets-test.js

```diff
@@ -0,0 +1,172 @@
+  if (options.failRight) {
+    tapTargetToTheRight.clientRects[0].left -= overlapAmount;
+    tapTargetToTheRight.clientRects[0].right -= overlapAmount;
+  }
+  if (options.failBelow) {
+    tapTargetBelow.clientRects[0].top -= overlapAmount;
+    tapTargetBelow.clientRects[0].bottom -= overlapAmount;
+  }
+  if (options.addFullyContainedTapTarget) {
+    targets.push({
+      snippet: '<contained></contained>',
+      clientRects: [
+        makeClientRects({
+          x: 0,
+          y: 0,
+        }),
+      ],
+    });
+  }
+  if (options.failSecondClientRect) {
+    mainTapTarget.clientRects.push(
+      makeClientRects({
+        x: 0,
+        y: overlapAmount,
+      })
+    );
+  }
+
+  return targets;
+}
+
+describe('SEO: Tap targets audit', () => {
+  it('passes when there are no tap targets', () => {
+    const auditResult = auditTapTargets([]);
+    assert.equal(auditResult.rawValue, true);
+    assert.equal(auditResult.score, 1);
+  });
+
+  it('passes when tap targets don\'t overlap', () => {
+    const auditResult = auditTapTargets(getBorderlineTapTargets());
+    assert.equal(auditResult.rawValue, true);
+  });
+
+  it('passes when a target is fully contained in an overlapping target', () => {
+    const auditResult = auditTapTargets(getBorderlineTapTargets({
+      addFullyContainedTapTarget: true,
+    }));
+    assert.equal(auditResult.rawValue, true);
+  });
+
+  it('fails if two tap targets overlaps each other horizontally', () => {
+    const auditResult = auditTapTargets(
+      getBorderlineTapTargets({
+        failRight: true,
+      })
+    );
+    assert.equal(auditResult.rawValue, false);
+    assert.equal(Math.round(auditResult.score * 100), 33);
+    const failure = auditResult.details.items[0];
+    assert.equal(failure.tapTarget.snippet, '<main></main>');
+    assert.equal(failure.overlappingTarget.snippet, '<right></right>');
+    assert.equal(failure.size, '10x10');
+    // Includes data for debugging/adjusting the scoring logic later on
+    assert.equal(failure.tapTargetScore, tapTargetSize * tapTargetSize);
+    assert.equal(failure.overlappingTargetScore, tapTargetSize * pxOverlappedByFinger);
+    assert.equal(failure.overlapScoreRatio, 0.3);
+    assert.equal(failure.width, 10);
+    assert.equal(failure.height, 10);
+  });
+
+  it('fails if a tap target overlaps vertically', () => {
+    const auditResult = auditTapTargets(
+      getBorderlineTapTargets({
+        failBelow: true,
+      })
+    );
+    assert.equal(auditResult.rawValue, false);
+  });
+
+  it('fails when one of the client rects overlaps', () => {
+    const auditResult = auditTapTargets(
+      getBorderlineTapTargets({
+        failSecondClientRect: true,
+      })
+    );
+    assert.equal(auditResult.rawValue, false);
+  });
+
+  it('reports 4 items if the main target is overlapped both vertically and horizontally', () => {
+    // Main is overlapped by right + below, right and below are each overlapped by main
```

Comment:
It's probably not a good sign that I've spent so much time with this audit by now and don't know the answer to this question yet buuuuuuuuut 😄 

Isn't this usually going to be the case? If it's symmetrical should we only report one of the directions?

Seems like the cases where it is symmetrical there's not necessarily an obvious winner of who is overlapping who and it's really just one failure. Sincere apologies if we've already covered this and reached this conclusion!

Reply:
Thanks for bringing this up – there definitely is a lot of duplication:

![screenshot 2019-01-04 at 09 51 40](https://user-images.githubusercontent.com/1303660/50682319-5dbc9900-1006-11e9-8f70-5e802644e0b8.png)

Possible issues with merging them:
- Fixing a problem will sometimes just result in the table element being flipped (I think that's fine)
- Users won't know if the overlapping target is also failing – but it probably doesn't matter to them (Maybe they are fine with one target being hard to tap but not the other? Maybe they'd like to know that fixing some failures will have more impact on their score?)
- The result JSON won't actually say what tap targets are failing, but if we or a third party wanted to visualize the results that would be good to know. Maybe there should be a `bidrectionaFailure: true` property?

Upsides:
- Less visual noise 
- Cuts tap targets JSON size in half in many cases

Sounds like a great idea to me.

@rviscomi Any thoughts on this?

Reply:
SGTM

Reply:
@rviscomi Right now you can't tell from the audit result which elements are failing. It could be that just the `tapTarget` is failing, or both the `tapTarget` and the `overlappingTarget` might be failing. Are we fine with that, or do we need to indicate symmetrical failures in the json somehow?

Reply:
I'm fine with the deduplication, even with the ambiguity it may introduce.

If tap targets are overlapping I think the natural response is to move them further apart, which solves each underlying failure, as opposed to increasing the dimensions of one target or another.

So let's launch with the simplified results, follow user feedback closely to see if there's any confusion, and revisit the config accordingly.